### DISCUSSION TITLE
Stanford CS144: Computer Network - CS自学指南

### DISCUSSION THREAD

**Discussion by giscus** [PKUFlyingPig/cs-self-learning#105] [Announcements]
# Stanford CS144: Computer Network - CS自学指南

CS自学指南

https://csdiy.wiki/%E8%AE%A1%E7%AE%97%E6%9C%BA%E7%BD%91%E7%BB%9C/CS144/

**Comment by PorunC** [comment]
这是我在学这门CS144路上翻阅过的博客或资料，把他们罗列在这里，希望让后来者提供一个参考，也希望自己能坚持做完吧～
## 参考资料

- [Lexssama's Blogs](https://lexssama.github.io/tags/CS144/)
- [huangrt01](https://github.com/huangrt01/CS-Notes/blob/master/Notes/Output/Computer-Networking-Lab-CS144-Stanford.md)
- [kiprey](https://kiprey.github.io/tags/CS144/)
- [康宇PL's Blog](https://www.cnblogs.com/kangyupl/p/stanford_cs144_labs.html)
- [doraemonzzz](http://doraemonzzz.com/tags/CS144/)
- [ViXbob's libsponge](https://vixbob.moe/25.html)
- [吃着土豆坐地铁的博客](https://www.epis2048.net/categories/Code/Stanford-CS144/)
- [Smith](https://www.inlighting.org/archives/2021-cs144-notes/#Lab-4-the-TCP-connection)
- [PKUFlyingPig](https://github.com/PKUFlyingPig/CS144-Computer-Network)
- [星遥见](https://www.cnblogs.com/weijunji/tag/CS144/)
- [EIMadrigal](https://www.cnblogs.com/EIMadrigal/p/15500472.html)
- [Joey](http://yuzijun.life/2021-02/CS144)

**Reply by IndifferentArea** [reply]
如果您不嫌麻烦地话或许可以pr一波直接放进正文里

**Reply by Cpt11** [reply]
推荐俩个便宜又好用的机场翻墙  平时如果只要google github 这些完全够用了
[一元机场](https://xn--4gq62f52gdss.com/#/register?code=APxjcVZh)   19元2年  100G/月

 [八方云](http://bafangyun.vip/#/register?code=0mobpNZJ)  12元/年  200G/月

**Reply by LingkKang** [reply]
好多链接都挂了 :(  ...

**Comment by PorunC** [comment]
这是我在学这门CS144路上翻阅过的博客或资料，把他们罗列在这里，希望让后来者提供一个参考，也希望自己能坚持做完吧～
## 参考资料

- [Lexssama's Blogs](https://lexssama.github.io/tags/CS144/)
- [huangrt01](https://github.com/huangrt01/CS-Notes/blob/master/Notes/Output/Computer-Networking-Lab-CS144-Stanford.md)
- [kiprey](https://kiprey.github.io/tags/CS144/)
- [康宇PL's Blog](https://www.cnblogs.com/kangyupl/p/stanford_cs144_labs.html)
- [doraemonzzz](http://doraemonzzz.com/tags/CS144/)
- [ViXbob's libsponge](https://vixbob.moe/25.html)
- [吃着土豆坐地铁的博客](https://www.epis2048.net/categories/Code/Stanford-CS144/)
- [Smith](https://www.inlighting.org/archives/2021-cs144-notes/#Lab-4-the-TCP-connection)
- [PKUFlyingPig](https://github.com/PKUFlyingPig/CS144-Computer-Network)
- [星遥见](https://www.cnblogs.com/weijunji/tag/CS144/)
- [EIMadrigal](https://www.cnblogs.com/EIMadrigal/p/15500472.html)
- [Joey](http://yuzijun.life/2021-02/CS144)

**Reply by eatvector** [reply]
lab4是真被恶心到了，看来半天一个bug也改不对，不过我还是要坚持住不看别人的代码。

**Comment by derecknowayback** [comment]
找队友

**Comment by taylover2016** [comment]
2022年2月到今天终于刷完了所有的课程以及Lab，纪念一下

**Comment by lightcream** [comment]
我是24考研的学生，408相关内容看以上相关课程可以满足应试需求吗，评论区如果有前辈看到的话，求解答一下

**Reply by derecknowayback** [reply]
考研看王道别看这些。

**Reply by lightcream** [reply]
感谢告知

**Comment by poxerial** [comment]
这个lab我个人感觉不是特别的好：实现的这个tcp协议栈纯粹就是照着要求做，什么超时重传的策略都在实验文档里写的清清楚楚了，基本上没有自己独立设计思考的空间；最难受的是很多细节实验文档里根本没说清楚，不得不面向测试用例编程。
我觉得只要实现一个符合协议并且性能可以接受的tcp就行了，为什么要把几乎所有东西都定死？

**Reply by waker-he** [reply]
同感，做下来感觉lab设计得不太行

**Reply by Schnabel-8** [reply]
我还在写check1，已经有这种感觉了，文档里有很多东西没写清楚，感觉细节可能在我们看不到的intro video里

**Reply by LiuChang-ao** [reply]
事实上这个lab已经简化很多了，如果真要“符合协议”，就要全按照RFC文档来写，比这个还抓狂。底层的东西都是这样，脏活累活多于设计和优化的部分。
不过我是真的佩服他们教授和助教，写了那么多的单元测试用例

**Reply by ghost** [reply]
lab0里看到的
We expect that you’ll benefit from attending the lab sessions and asking clarifying questions if you find something to be ambiguous and you think the answer matters。

**Reply by ghost** [reply]
The lab documents aren’t “specifications”—meaning they’re not intended to be consumed in a one-way fashion. They’re written closer to the level of detail that a software
engineer will get from a boss or client. We expect that you’ll benefit from attending the
lab sessions and asking clarifying questions if you find something to be ambiguous and
you think the answer matters. We’ll update the “lab FAQ” document on the course
website in response to late questions that need clarification.

**Reply by chenfs12138** [reply]
额，这虽然是跟开发很接近，但是总是课程，需要一个评分标准，而且协议就这样，UB一大堆，c的标准都多少年了各种编译器还有不完全一致的行为和各种专属特性，最主要的是不定死最后搭起来的不一定能用啊，能用才是王道

**Comment by zevin02** [comment]
cs144这门课有群聊一起学习的吗

**Reply by eatvector** [reply]
之前做到lab4被bug卡住了，现在又有时间准备继续做一下，同求群。

**Reply by JhoniWu** [reply]
同求同求，想自学一下计网

**Reply by eatvector** [reply]
终于把lab4做完了，被bug卡了十二三天，要疯了。

**Comment by SonglinLife** [comment]
2023.3.3我也完成了这个lab，Mark下，感谢博主让我知道了这样的课程。

我的代码在： [songlinlife](https://github.com/SonglinLife/tcpImplementation)

我创建了一个群： 751581942， 欢迎来玩啊

**Comment by dengyunsheng250** [comment]
找队友

**Reply by yao-wu-yang** [reply]
一起做的队友 微信：15769415739

**Comment by XueSongTap** [comment]
今天看好像更新到2023课程了sponge仓库删了，哪里能找到原来的sponge仓库啊

**Reply by amcones** [reply]
可以fork之前别人写完的仓库然后回退提交，btw为啥不写新的呢

**Reply by fxy0278** [reply]
请问在哪可以看到这门课之前的实验文档，比如2021年秋季学期的

**Reply by amcones** [reply]
可以在一些大佬的博客里找找，翻译版好像还是挺好找的

**Comment by xinyangcui** [comment]
有大佬件建群一起做吗

**Reply by HangX-Ma** [reply]
推荐做21年的sponge版本，23年的做了发现 connection 这部分 （据说最难的给删掉了），然后 lab6 和 lab7 就全是 optional 不用写代码了。楼上的那个 751581942 群现在还有活跃的人，可以一起。

**Reply by ThreadEater** [reply]
想问下各位大佬是用哪种方式配的环境啊？

**Comment by ThreadEater** [comment]
想问下各位大佬是用哪种方式配的环境啊？
1. On Intel/AMD computers: Use a VM image that we prepared in VirtualBox
2. Use a Google Cloud virtual machine using our class's coupon code
3. Use your own GNU/Linux installation with a C++20 compiler (GCC 12 or later, clang 15 or later)
4. On ARM MacBooks and Macs: please install the UTM virtual machine software and use our ARM64 GNU/Linux virtual machine image
我是win+Intel，装了wsl

**Reply by FlipThatCover** [reply]
我在硬盘上开了500G，装了ubuntu 22.10。因为clang编译比GCC快太多所以选的clang
clang官网没找到适用于ubuntu的版本（也可能是我太菜了不会设置），最后自己下载源代码编译了一个

**Reply by werner-balonius** [reply]
我是第四种，环境还是蛮好配的

**Comment by FlipThatCover** [comment]
今天写完了Lab 1，一顿操作throughput优化到了13.22Gbits，兴奋之余给几点tips
1. 不要用std::map<uint64_t, std::string>，搜索区间顺而想到二叉树的想法没错，但std::map的insert和erase比较慢，尤其是erase，std::map适合海量数据下的查找，而非频繁的插入删除（如果想锻炼code能力的当我没说）
2. 最容易想到的方法之一是std::unordered_map<uint64_t, char> 然而哈希表的写入和擦除速度同样感人，实测速度为0.3Gbits。这个方法唯一的好处就是写起来简单，调试简单。我当初写完就30多行。
3. 用好std::move
4. 鉴于本实验恶心的边界条件，不建议在初期就对代码做逻辑上的优化
5. （暗示）数据量不到50万的话，顺序结构也是很快的

**Comment by wuyafei** [comment]
用chatGPT翻译了一个版本，效果比普通机翻好很多，喜欢中英双语的可以看看

**Reply by wuyafei** [reply]
https://www.bilibili.com/video/BV16o4y1V7RQ/?vd_source=2b9f6fcc0b53b5f5e1607a9e3e8c83bd

**Comment by LRL52** [comment]
2023.8.19 我也完成了 CS144 Lab，使用的是 21 年的 sponge 版本。这是我所有 Lab0-7 的个人笔记：https://lrl52.top/category/cs144/   希望对后人有所帮助鸭~😀

**Reply by Lengdrich** [reply]
感谢

**Comment by shiyang1026** [comment]
有做Spring 23的伙伴儿吗？
lab0 在 cmake --build build 时候
报错 ...tests/common.hh:16:32:error: invalid return type ‘std::string’ {aka ‘std::__cxx11::basic_string<char>’} of ‘constexpr’ function ‘static constexpr std::string ExpectationViolation::boolstr(bool)’
有人遇到吗

**Reply by HangX-Ma** [reply]
检查一下系统版本以及安装的依赖库版本，需要用指导书推荐的。

**Reply by shiyang1026** [reply]
感谢，已经解决, 是Ubuntu版本的问题，升级到实验要求的version就好了

**Reply by zchern** [reply]
我在做Winter 24时也遇到了这个问题

**Reply by zchern** [reply]
关于这个问题我的经验总结1：ubuntu版本问题，22.10版本及以上，但22.10是个历史版本，可以使用22.4更新到23.4或者直接使用23.4，值得注意的是。清华镜像的23.4ios有问题，是用不了。2：gcc版本问题，需要gcc 11及其以上版本才能支持c++20标准的编程

**Comment by habaohaba** [comment]
想问这门课的网站和proj代码是不公开了吗？

**Reply by fei0319** [reply]
就在本月初因不明原因全部下架了

**Reply by habaohaba** [reply]
不知道教授未来是什么计划，是决定以后都不公开了或者只是更新内容。唉，遗憾。
确实网上不尊重规定，公开solution代码甚至盗用资源，售卖资源的人太多了，教授也没理由公开这些资源了。

**Reply by goahead656** [reply]
希望还能继续开源趴，555

**Reply by okayokayok** [reply]
想做lab，看来是寄了

**Reply by goahead656** [reply]
可以看看参考资料学一学

**Reply by mabinhang2021** [reply]
重新开放了

**Reply by zchern** [reply]
24 Winter开放了

**Comment by Stone-face** [comment]
请问哪里还能找到lab的初始代码呢？官方的好像已经不公开了。万分感谢~

**Reply by mabinhang2021** [reply]
您可以看我的评论，以及重新开放了

**Comment by mabinhang2021** [comment]
CS144 2024 WINTER 重新开放了代码。我刚完成lab0并且写了一篇文章放在知乎上，链接是CS144 2024 WINTER Lab Checkpoint 0: networking warmup - 加德满都的风铃的文章 - 知乎 https://zhuanlan.zhihu.com/p/677623671
期待大家拉我建群一起讨论，qq：1223841837

**Reply by zchern** [reply]
才开始lab0，已经因为环境焦头烂额额。。。

**Reply by mabinhang2021** [reply]
加油，我也刚刚做完lab1

**Comment by hychen11** [comment]
想问问做哪年的版本比较好呢？minnow还是sponge呢？

**Reply by zchern** [reply]
只有minnow可以做了

**Comment by Axlashh** [comment]
熬了几天夜终于把lab4写完了，简单记录了一下lab4的debug过程，放在我仓库的[README](https://github.com/Axlashh/CS144/blob/main/README.md)里了，没有大佬们的博客那么详细，希望能帮助到一些人。

**Reply by thewangcj** [reply]
被lab4折磨了三天，我也是一个一个用例单独跑的，整的我怀疑人生，还以为自己的理解有问题，为什么大家的代码都这么简单，看了你这笔记才知道原来用的是 relaxed 版本啊，具体命令在 tests.cmake ：
add_custom_target (check_lab4 COMMAND "${PROJECT_SOURCE_DIR}/tun.sh" check 144 145
                              COMMAND ${CMAKE_CTEST_COMMAND} --output-on-failure --timeout 10 -R "^t_"
                              COMMENT "Testing libsponge...")
只会执行 t_ 开头的测试用例，我真的……

**Comment by CauchyJenson** [comment]
lab4给我一种无从下手的感觉。。。。。。

**Comment by zhangeecs** [comment]
我发现油管的视频说他自己是从B站搬的；
只是B站网址：【英字【计算机网络导论】斯坦福大学 Introduction to Computer Networking (CS 144)】 https://www.bilibili.com/video/BV1wt41167iN

**Comment by zhangeecs** [comment]
EIMadrigal的博客搬家了，https://eimadrigal.github.io/posts/cs144-lab/
Joey的博客我这边打不开，不知道其他人能不能打开

**Comment by coderbaibai** [comment]
感觉这个lab不考察对整个structure的理解，仅仅只是一些算法的实现（有测试用例的话比Leetcode还简单）。要想有收获还是得深入到它的整个源码来思考架构和原理。

**Comment by QWenQ** [comment]
2024WINETER 把Lab4：TCPConnection删了，换成了“Lab4：interoperating in the real world”，我就说看着上面的评论，怎么感觉跟自己的经历有点格格不入呢......

**Comment by jiuyueshenhua** [comment]
最近建的一个QQ交流群：642596152
欢迎同志者交流。

**Comment by dghtucs** [comment]
QQ群694720782讨论交流

**Comment by puji4810** [comment]
课程似乎再次关闭了？看了一眼评论区，24winter是1月初开放的，希望过些天重新开放吧，本来准备寒假学的

**Reply by puji4810** [reply]
前些天已经重新开了

**Reply by RisingUppercut** [reply]
请问24winter的课程网址是啥啊

**Comment by curbe454** [comment]
好稀奇, 这门课居然是直接部署在github上的静态页面, 没有使用Jekyll, 直接git clone一波全部下载了
其他有些课程找资源真的找得我累了...

**Comment by Oooscar8** [comment]
2025.2.27 挑战全网第一个（划掉）写完CS144 2025 winter的Lab 0-6，这里是我的Lab笔记（除了Lab4，因为Lab4不用coding）：https://oooscar8.github.io/categories/Computer-Networking/CS144/ 自认为还是挺详细的，重点梳理了实现思路和容易出错的点，希望对你有帮助！

**Reply by RisingUppercut** [reply]
nb，你现在大几啊

**Reply by sleepy-shawn** [reply]
牛的
遇到不会的就来看你的code

**Reply by Monsteer714** [reply]
强，现在做到lab2，你的笔记很有帮助

**Comment by Albert-tru** [comment]
2025有小伙伴一起做嘛！！！

**Reply by sleepy-shawn** [reply]
有的 有的 刚学完王道计网 感觉不写代码太浅薄了 所以来学学144

**Reply by Albert-tru** [reply]
我也是正在看计算机网络自顶向下那本书，不嫌弃的话，可以交流一下吗 qq2556066885

**Comment by Albert-tru** [comment]
唉GG 我的ubuntu里怎么上不了github了，梯子试了好几个教程也装不上，GGGGGG，qaq

**Reply by FeiNiaoBF** [reply]
你是用虚拟机的吗？可以问问ai试试

**Comment by L1w-Y** [comment]
25年4.14开始做这个lab，欢迎正在做的朋友一起交流，qq1378615481

**Comment by liliaoyuan** [comment]
还有最近在看144的老哥吗，交流qq：3202835415@qq.com

另外还有sponge版本的资源，找个这个直接build就报错了，求个分享

**Comment by HT4w5** [comment]
最近 Winter 2025 的 starter code 仓库突然不公开了，这里有[完整备份](https://github.com/HT4w5/minnow-winter-2025)。
另外希望各位尊重作者的要求，不要把自己的答案公开发表。

**Reply by puji4810** [reply]
这两年都是这样，开几个月就关了，建议还是快点做，别一直拖着。不然就得等过年开2026 winter了

**Comment by IMElTgp** [comment]
现在是不是没法做了啊。。做完checkpoint0告诉我需要git merge 但是官方的repo好像设private了😭

**Reply by yanlu788** [reply]
兄弟我也刚开始做能交流一下吗？我q2927637779

**Reply by IMElTgp** [reply]
我打算烂尾等win2026了。。😂

**Reply by HT4w5** [reply]
你看看楼下我发的有。

**Reply by Keedled** [reply]
可以看看别人的仓库，保存有之前的版本

**Reply by hiddenblue** [reply]
其实是因为9月份他们要开冬季学期了，所以有那么几天会更新仓库，所以就没法正常merge，过一周就好了。

**Comment by Zlatanwic** [comment]
大佬们问下，ucb cs168和stf cs144这两个综合来讲比较推荐哪个呢

**Reply by Keedled** [reply]
我写完了CS144，个人觉得实验文档写的不太好，很多得面向样例编程。

**Comment by coderhxjin** [comment]
https://coderhxjin.github.io/cs144-2021-backup.github.io/

**Reply by coderhxjin** [reply]
找了一个21年版网站的仓库，自己 fork 了一份

**Comment by rinevard** [comment]
我在 CS144 和 CS168 间犹豫了很久。前者视频资料比较古早但作业用了现代 C++（我一直想学现代 C++），后者学习资料丰富、最新视频公开，但作业用的是 Python。最后我 DIY 了一个 CS144 和 CS168 的混合路线，学 CS168 的课但做 CS144 的作业，实践下来非常可行，满足了我同时学习计网和现代 C++ 的需求。

我把 DIY 的学习日程表、CS144 Fall 2025 的 Starter Code、实验文档及其机翻中英对照备份在了 [NetworkDIY](https://github.com/rinevard/NetworkDIY) 中，希望对你有帮助。
